File: examples/cpp/Misc/MultiDevice/multi_device_m8_fsync.cpp
#include <algorithm>
#include <chrono>
#include <csignal>
#include <cstdlib>
#include <iostream>
#include <memory>
#include <opencv2/core.hpp>
#include <opencv2/core/mat.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/opencv.hpp>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

#include "depthai/capabilities/ImgFrameCapability.hpp"
#include "depthai/common/CameraBoardSocket.hpp"
#include "depthai/common/CameraExposureOffset.hpp"
#include "depthai/common/M8FsyncRoles.hpp"
#include "depthai/depthai.hpp"
#include "depthai/pipeline/MessageQueue.hpp"
#include "depthai/pipeline/Node.hpp"
#include "depthai/pipeline/datatype/ImgFrame.hpp"
#include "depthai/pipeline/node/Camera.hpp"
#include "depthai/pipeline/node/Sync.hpp"
#include "depthai/xlink/XLinkConnection.hpp"

class FPSCounter {
   public:
    std::vector<std::chrono::time_point<std::chrono::steady_clock>> frameTimes;

    void tick() {
        auto now = std::chrono::steady_clock::now();
        frameTimes.push_back(now);
        if(frameTimes.size() > 100) {
            frameTimes.erase(frameTimes.begin(), frameTimes.begin() + (frameTimes.size() - 100));
        }
    }

    float getFps() {
        if(frameTimes.size() <= 1) {
            return 0.0f;
        }
        // Calculate the FPS
        return float(frameTimes.size() - 1) * 1e6 / std::chrono::duration_cast<std::chrono::microseconds>(frameTimes.back() - frameTimes.front()).count();
    }
};

dai::Node::Output* createPipeline(dai::Pipeline& pipeline, dai::CameraBoardSocket socket, int sensorFps, dai::M8FsyncRole role) {
    std::shared_ptr<dai::node::Camera> cam;
    if(role == dai::M8FsyncRole::MASTER) {
        cam = pipeline.create<dai::node::Camera>()->build(socket, std::nullopt, sensorFps);
    } else {
        cam = pipeline.create<dai::node::Camera>()->build(socket, std::nullopt);
    }

    auto output = cam->requestOutput(std::make_pair(640, 480), dai::ImgFrame::Type::NV12, dai::ImgResizeMode::STRETCH);
    return output;
}

namespace {
bool running = true;
}

void interruptHandler(int sig) {
    if(running) {
        std::cout << "Interrupted! Exiting..." << std::endl;
        running = false;
    } else {
        std::cout << "Exiting now!" << std::endl;
        exit(0);
    }
}

int main(int argc, char** argv) {
    if(argc < 5) {
        std::cout << "Usage: " << argv[0]
                  << " <target_fps> <recv-all-timeout-sec> <sync-threshold-sec> <initial-sync-timeout-sec> [<device_ip_1> <device_ip_2> [device_ip_3]] ..."
                  << std::endl;
        std::exit(0);
    }

    signal(SIGINT, interruptHandler);

    float targetFps = std::stof(argv[1]);

    int recvAllTimeoutSec = std::stoi(argv[2]);
    float syncThresholdSec = std::stof(argv[3]);
    int initialSyncTimeoutSec = std::stoi(argv[4]);

    std::vector<dai::DeviceInfo> deviceInfos;

    if(argc > 5) {
        for(int i = 5; i < argc; i++) {
            deviceInfos.emplace_back(std::string(argv[i]));
        }
    } else {
        deviceInfos = dai::Device::getAllAvailableDevices();
    }

    if(deviceInfos.size() < 2) {
        std::cout << "At least two devices are required for this example." << std::endl;
        std::exit(0);
    }

    std::vector<dai::Node::Output*> masterNodes;
    std::vector<std::shared_ptr<dai::MessageQueue>> slaveQueues;
    std::vector<dai::Pipeline> masterPipelines;
    std::vector<dai::Pipeline> slavePipelines;
    std::vector<std::string> deviceIds;

    std::vector<std::shared_ptr<dai::InputQueue>> inputQueues;
    std::vector<std::string> outputNames;

    for(auto deviceInfo : deviceInfos) {
        auto pipeline = dai::Pipeline(std::make_shared<dai::Device>(deviceInfo));
        auto device = pipeline.getDefaultDevice();

        auto role = device->getM8FsyncRole();

        std::cout << "=== Connected to " << deviceInfo.getDeviceId() << std::endl;
        std::cout << "    Device ID: " << device->getDeviceId() << std::endl;
        std::cout << "    Num of cameras: " << device->getConnectedCameras().size() << std::endl;

        auto socket = device->getConnectedCameras()[0];
        if(device->getProductName().find("OAK4-D") != std::string::npos || device->getProductName().find("OAK-4-D") != std::string::npos) {
            socket = device->getConnectedCameras()[1];
        }

        auto outNode = createPipeline(pipeline, socket, targetFps, role);

        if(role == dai::M8FsyncRole::MASTER) {
            device->setM8StrobeEnable(true);
            masterPipelines.push_back(pipeline);
            masterNodes.push_back(outNode);
            std::cout << device->getDeviceId() << " is master" << std::endl;
        } else if(role == dai::M8FsyncRole::SLAVE) {
            slavePipelines.push_back(pipeline);
            slaveQueues.push_back(outNode->createOutputQueue());
            std::cout << device->getDeviceId() << " is slave" << std::endl;
        } else {
            throw std::runtime_error("Don't know how to handle role " + dai::toString(role));
        }

        deviceIds.push_back(deviceInfo.getXLinkDeviceDesc().name);
    }

    if(masterPipelines.size() > 1) {
        throw std::runtime_error("Multiple masters detected!");
    }
    if(masterPipelines.size() == 0) {
        throw std::runtime_error("No master detected!");
    }
    if(slavePipelines.size() < 1) {
        throw std::runtime_error("No slaves detected!");
    }

    auto sync = masterPipelines[0].create<dai::node::Sync>();
    sync->setRunOnHost(true);
    sync->setSyncThreshold(std::chrono::nanoseconds(long(round(1e9 * 0.5f / targetFps))));
    masterNodes[0]->link(sync->inputs["master"]);
    outputNames.push_back("master");

    for(unsigned long i = 0; i < slaveQueues.size(); i++) {
        auto name = std::string("slave_") + std::to_string(i);
        outputNames.push_back(name);
        auto input_queue = sync->inputs[name].createInputQueue();
        inputQueues.push_back(input_queue);
    }

    auto queue = sync->out.createOutputQueue();

    for(auto p : masterPipelines) {
        p.start();
    }
    for(auto p : slavePipelines) {
        p.start();
    }

    FPSCounter fpsCounter;

    std::optional<std::shared_ptr<dai::MessageGroup>> latestFrameGroup;
    bool firstReceived = false;
    auto startTime = std::chrono::steady_clock::now();
    auto prevReceived = std::chrono::steady_clock::now();

    std::optional<std::chrono::time_point<std::chrono::steady_clock>> initialSyncTime;

    bool waitingForInitialSync = true;

    while(running) {
        for(int i = 0; i < slaveQueues.size(); i++) {
            while(slaveQueues[i]->has()) {
                inputQueues[i]->send(slaveQueues[i]->get());
            }
        }

        while(queue->has()) {
            auto syncData = queue->get();
            latestFrameGroup = std::dynamic_pointer_cast<dai::MessageGroup>(syncData);
            if(!firstReceived) {
                firstReceived = true;
                initialSyncTime = std::chrono::steady_clock::now();
            }
            prevReceived = std::chrono::steady_clock::now();
            fpsCounter.tick();
        }

        if(!firstReceived) {
            auto endTime = std::chrono::steady_clock::now();
            auto elapsedSec = std::chrono::duration_cast<std::chrono::seconds>(endTime - startTime).count();
            if(elapsedSec >= recvAllTimeoutSec) {
                std::cout << "Timeout: Didn't receive all frames in time: " << elapsedSec << std::endl;
                running = false;
            }
        }

        if(latestFrameGroup.has_value() && size_t(latestFrameGroup.value()->getNumMessages()) == outputNames.size()) {
            std::vector<std::chrono::time_point<std::chrono::steady_clock>> tsValues;
            for(auto name : outputNames) {
                auto frame = latestFrameGroup.value()->get<dai::ImgFrame>(name);
                tsValues.push_back(frame->getTimestamp(dai::CameraExposureOffset::END));
            }

            cv::Mat imgs;
            for(int i = 0; i < outputNames.size(); i++) {
                auto msg = latestFrameGroup.value()->get<dai::ImgFrame>(outputNames[i]);
                auto fps = fpsCounter.getFps();
                auto frame = msg->getCvFrame();

                cv::putText(frame, deviceIds[i] + " (" + outputNames[i] + ")", {20, 40}, cv::FONT_HERSHEY_SIMPLEX, 0.6, {0, 127, 255}, 2, cv::LINE_AA);

                cv::putText(frame,
                            "Timestamp: " + std::to_string(1e-6 * std::chrono::duration_cast<std::chrono::microseconds>(tsValues[i].time_since_epoch()).count())
                                + " | FPS: " + std::to_string(fps).substr(0, 5),
                            {20, 80},
                            cv::FONT_HERSHEY_SIMPLEX,
                            0.6,
                            {255, 0, 50},
                            2,
                            cv::LINE_AA);

                if(i == 0) {
                    imgs = frame;
                } else {
                    cv::hconcat(frame, imgs, imgs);
                }
            }

            auto delta = *std::max_element(tsValues.begin(), tsValues.end()) - *std::min_element(tsValues.begin(), tsValues.end());

            bool syncStatus = std::chrono::duration_cast<std::chrono::seconds>(delta).count() < syncThresholdSec;
            std::string syncStatusStr = (syncStatus) ? "in sync" : "out of sync";

            cv::Scalar color = (syncStatus) ? cv::Scalar(0, 255, 0) : cv::Scalar(0, 0, 255);

            if(!syncStatus && waitingForInitialSync) {
                auto endTime = std::chrono::steady_clock::now();
                auto elapsedSec = std::chrono::duration_cast<std::chrono::seconds>(endTime - initialSyncTime.value()).count();
                if(elapsedSec >= initialSyncTimeoutSec) {
                    std::cout << "Timeout: Didn't sync frames in time" << std::endl;
                    running = false;
                }
            }

            if(syncStatus && waitingForInitialSync) {
                std::cout << "Sync status: " << syncStatusStr << std::endl;
                waitingForInitialSync = false;
            }

            if(!syncStatus && !waitingForInitialSync) {
                std::cout << "Sync error: Sync lost, threshold exceeded " << std::chrono::duration_cast<std::chrono::microseconds>(delta).count() << " us"
                          << std::endl;
                running = false;
            }

            cv::putText(imgs,
                        syncStatusStr + " | delta = "
                            + std::to_string(1e-3 * float(std::chrono::duration_cast<std::chrono::microseconds>(delta).count())).substr(0, 5) + " ms",
                        {20, 120},
                        cv::FONT_HERSHEY_SIMPLEX,
                        0.7,
                        color,
                        2,
                        cv::LINE_AA);

            cv::imshow("synced_view", imgs);

            latestFrameGroup.reset();
        }

        if(cv::waitKey(1) == 'q') {
            break;
        }
    }

    return 0;
}

File: examples/python/Misc/MultiDevice/multi_device_m8_fsync.py
#!/usr/bin/env python3

"""
Minimal changes to original script:
  * Adds simple timestamp-based synchronisation across multiple devices.
  * Presents frames side‑by‑side when they are within 1 / FPS seconds.
  * Keeps v3 API usage and overall code structure intact.
"""

import contextlib
import datetime

import cv2
import depthai as dai
import time
import math

import argparse
import signal
import numpy as np
# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------
SET_MANUAL_EXPOSURE = False  # Set to True to use manual exposure settings
# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------
class FPSCounter:
    def __init__(self):
        self.frameTimes = []

    def tick(self):
        now = time.time()
        self.frameTimes.append(now)
        self.frameTimes = self.frameTimes[-100:]

    def getFps(self):
        if len(self.frameTimes) <= 1:
            return 0
        # Calculate the FPS
        return (len(self.frameTimes) - 1) / (self.frameTimes[-1] - self.frameTimes[0])


# ---------------------------------------------------------------------------
# Pipeline creation (unchanged API – only uses TARGET_FPS constant)
# ---------------------------------------------------------------------------
def createPipeline(pipeline: dai.Pipeline, socket: dai.CameraBoardSocket, sensorFps: int, role: dai.M8FsyncRole):
    cam = None
    if role == dai.M8FsyncRole.MASTER:
        cam = (
            pipeline.create(dai.node.Camera)
            .build(socket, sensorFps=sensorFps)
        )
    else:
        cam = (
            pipeline.create(dai.node.Camera)
            .build(socket)
        )

    output = (
        cam.requestOutput(
            (640, 480), dai.ImgFrame.Type.NV12, dai.ImgResizeMode.STRETCH
        )
    )
    return pipeline, output

running = True

def interruptHandler(sig, frame):
    global running
    if running:
        print("Interrupted! Exiting...")
        running = False
    else:
        print("Exiting now!")
        exit(0)

signal.signal(signal.SIGINT, interruptHandler)

parser = argparse.ArgumentParser(add_help=False)
parser.add_argument("-f", "--fps", type=float, default=30.0, help="Target FPS", required=False)
parser.add_argument("-d", "--devices", default=[], nargs="+", help="Device IPs", required=False)
parser.add_argument("-t1", "--recv-all-timeout-sec", type=float, default=10, help="Timeout for receiving the first frame from all devices", required=False)
parser.add_argument("-t2", "--sync-threshold-sec", type=float, default=3e-3, help="Sync threshold in seconds", required=False)
parser.add_argument("-t3", "--initial-sync-timeout-sec", type=float, default=4, help="Timeout for synchronization to complete", required=False)
args = parser.parse_args()

if len(args.devices) == 0:
    deviceInfos = dai.Device.getAllAvailableDevices()
else:
    deviceInfos = [dai.DeviceInfo(ip) for ip in args.devices] #The master camera needs to be first here

assert len(deviceInfos) > 1, "At least two devices are required for this example."

targetFps = args.fps  # Must match sensorFps in createPipeline()
recvAllTimeoutSec = args.recv_all_timeout_sec

syncThresholdSec = args.sync_threshold_sec
initialSyncTimeoutSec = args.initial_sync_timeout_sec
# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
with contextlib.ExitStack() as stack:

    masterNodes = []
    slaveQueues = []
    slavePipelines = []
    masterPipelines = []
    deviceIds = []

    inputQueues = []
    slaveInputNames = []
    outputNames = []

    for idx, deviceInfo in enumerate(deviceInfos):
        pipeline = stack.enter_context(dai.Pipeline(dai.Device(deviceInfo)))
        device = pipeline.getDefaultDevice()

        role = device.getM8FsyncRole()

        print("=== Connected to", deviceInfo.getDeviceId())
        print("    Device ID:", device.getDeviceId())
        print("    Num of cameras:", len(device.getConnectedCameras()))

        socket = device.getConnectedCameras()[0]
        if "OAK4-D" in device.getProductName() or \
            "OAK-4-D" in device.getProductName():
            socket = device.getConnectedCameras()[1]

        pipeline, outNode = createPipeline(pipeline, socket, targetFps, role)

        if role == dai.M8FsyncRole.MASTER:
            device.setM8StrobeEnable(True)
            device.setM8StrobeLimits(0.05, 0.95)
            print(f"{device.getDeviceId()} is master")
            masterPipelines.append(pipeline)
            masterNodes.append(outNode)
        elif role == dai.M8FsyncRole.SLAVE:
            slavePipelines.append(pipeline)
            slaveQueues.append(outNode.createOutputQueue())
            print(f"{device.getDeviceId()} is slave")
        else:
            raise RuntimeError(f"Don't know how to handle role {role}")

        deviceIds.append(deviceInfo.getXLinkDeviceDesc().name)

    if len(masterPipelines) > 1:
        raise RuntimeError("Multiple masters detected!")

    if len(masterPipelines) == 0:
        raise RuntimeError("No master detected!")

    if len(slavePipelines) < 1:
        raise RuntimeError("No slaves detected!")

    sync = masterPipelines[0].create(dai.node.Sync)
    sync.setRunOnHost(True)
    sync.setSyncThreshold(datetime.timedelta(milliseconds=1000 / (2 * targetFps)))
    masterNodes[0].link(sync.inputs["master"])
    outputNames.append("master")

    for i in range(len(slaveQueues)):
        name = f"slave_{i}"
        slaveInputNames.append(name)
        outputNames.append(name)
        input_queue = sync.inputs[name].createInputQueue()
        inputQueues.append(input_queue)

    queue = sync.out.createOutputQueue()

    for p in masterPipelines:
        p.start()

    for p in slavePipelines:
        p.start()


    fpsCounter = FPSCounter()

    latestFrameGroup = None
    firstReceived = False
    startTime = datetime.datetime.now()
    prevReceived = datetime.datetime.now()

    initialSyncTime = None
    waitingForSync = True

    while running:

        for i, slq in enumerate(slaveQueues):
            while slq.has():
                inputQueues[i].send(slq.get())

        while queue.has():
            latestFrameGroup = queue.get()
            if not firstReceived:
                firstReceived = True
                initialSyncTime = datetime.datetime.now()
            prevReceived = datetime.datetime.now()
            fpsCounter.tick()

        if not firstReceived:
            endTime = datetime.datetime.now()
            elapsedSec = (endTime - startTime).total_seconds()
            if elapsedSec >= recvAllTimeoutSec:
                print(f"Timeout: Didn't receive all frames in time: {elapsedSec:.2f} sec")
                running = False

        # -------------------------------------------------------------------
        # Synchronise: we need at least one frame from every camera and their
        # timestamps must align within SYNC_THRESHOLD_SEC.
        # -------------------------------------------------------------------
        if latestFrameGroup is not None and latestFrameGroup.getNumMessages() == len(outputNames):
            tsValues = [latestFrameGroup[name].getTimestamp(dai.CameraExposureOffset.END).total_seconds() for name in outputNames]
            # Build composite image side‑by‑side
            imgs = []
            fps = fpsCounter.getFps()

            for i, outputName in enumerate(outputNames):
                msg = latestFrameGroup[outputName]
                frame = msg.getCvFrame()
                cv2.putText(
                    frame,
                    f"{deviceIds[i]} ({outputName})",
                    (20, 40),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    0.6,
                    (0, 127, 255),
                    2,
                    cv2.LINE_AA,
                )
                cv2.putText(
                    frame,
                    f"Timestamp: {tsValues[i]} | FPS:{fps:.2f}",
                    (20, 80),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    0.6,
                    (255, 0, 50),
                    2,
                    cv2.LINE_AA,
                )
                imgs.append(frame)

            delta = max(tsValues) - min(tsValues)

            syncStatus = abs(delta) < syncThresholdSec
            syncStatusStr = "in sync" if syncStatus else "out of sync"

            if not syncStatus and waitingForSync:
                endTime = datetime.datetime.now()
                elapsedSec = (endTime - initialSyncTime).total_seconds()
                if elapsedSec >= initialSyncTimeoutSec:
                    print("Timeout: Didn't sync frames in time")
                    running = False

            if syncStatus and waitingForSync:
                print(f"Sync status: {syncStatusStr}")
                waitingForSync = False

            if not syncStatus and not waitingForSync:
                print(f"Sync error: Sync lost, threshold exceeded {delta * 1e6} us")
                running = False

            color = (0, 255, 0) if syncStatusStr == "in sync" else (0, 0, 255)
            
            cv2.putText(
                imgs[0],
                f"{syncStatusStr} | delta = {delta*1e3:.3f} ms",
                (20, 120),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.7,
                color,
                2,
                cv2.LINE_AA,
            )

            cv2.imshow("synced_view", cv2.hconcat(imgs))

            latestFrameGroup = None  # Wait for next batch

        if cv2.waitKey(1) & 0xFF == ord("q"):
            break

cv2.destroyAllWindows()
